### PR TITLE
DEV-11830 - Stack feature - UI

### PR DESCRIPTION
**Pull request by mayhuberman**

### CODE REVIEW

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 1]
```diff
@@ -64,6 +80,23 @@ angular.module('apollo')
...
+        var isValueExist = function(array, value) {
+            for(var i=0, len=array.length; i<len; i++) {
+                if(array[i] == value) {
+                    return true;
+                }
+            }
+            return false;
+        };
+
+        $scope.showPopover = function() {
```
[the remark this answers is not in the document]
Comment (mayhuberman) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: deleted:)

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 2]
```diff
@@ -64,6 +80,23 @@ angular.module('apollo')
         var loadedGroupsEnvironmentId;
         var loadedGroupsServiceId;
 
+        var isValueExist = function(array, value) {
```
[the remark this answers is not in the document]
Comment (mayhuberman) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: I changed it to includes

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 3]
```diff
@@ -445,26 +532,115 @@ angular.module('apollo')
...
+                addStackToDisplayEnvironments(environmentsStack);
+
+                var environmentsNames = [];
+                environmentsStack.environments.forEach(function(environmentId) {
+                    var environment = allEnvironments.find(x => x.id === environmentId);
+                    var environmentName = environment.name;
+                    environmentsNames = environmentsNames.concat(environmentName);
+                });
+
+                $scope.environmentsInStack[environmentsStack.id] = environmentsNames;
```
[the remark this answers is not in the document]
Comment (mayhuberman) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: no, this array holds the names of the environments of every stack

File: ui/app/index.html [hunk 4]
```diff
@@ -75,11 +75,11 @@
 
         <!-- Custom Fonts -->
 
-        <script type="application/javascript">
-            window.alert = function(arguments) {
-                console.log("Caught alert:", arguments)
-            };
-        </script>
+        <!--<script type="application/javascript">-->
+            <!--window.alert = function(arguments) {-->
```
Comment (roiravhon) [inline ui/app/index.html]: Still needs revert

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 5]
```diff
@@ -124,41 +141,48 @@ angular.module('apollo')
...
-
-                }, function (error) {
-                    // End spinner
-                    usSpinnerService.stop('deployment-spinner');
-
-                    // 403 are handled generically on the interceptor
-                    if (error.status !== 403) {
-                        growl.error("Got from apollo API: " + error.status + " (" + error.statusText + ")", {ttl: 7000});
+                $scope.selectedEnvironments.forEach(function(environment) {
+                    $scope.selectedGroups.forEach(function(group) {
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: Lets aggregate per environment, and not sent a POST by group

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 6]
```diff
@@ -291,6 +322,16 @@ angular.module('apollo')
                 });
         };
 
+        $scope.getGroupsNamesPerEnvironment = function(environmentId) {
+             var groupsNamesPerEnvironment = [];
+             $scope.selectedGroups.forEach(function(group) {
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: Maybe convert to filter? just for readability. don't have to

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 7]
```diff
@@ -303,40 +344,168 @@ angular.module('apollo')
...
             else {
-                if ($scope.isGroupDeployment) {
-                    $scope.selectedEnvironments = [];
+                if(environment.isStack == true) {
+                    var stack = environment;
+                    var environmentsNames = [];
+                    $scope.selectedEnvironmentsAndStacks.push(stack);
+                    stack.environments.forEach(function(environmentId) {
+                        var index = $scope.selectedEnvironments.indexOf(environment);
+                        if (index > -1) {
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: If I select an environment, and then a stack that this environment is part of - won't this remove the selection for me?
There is no sense in this part IMO

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 8]
```diff
@@ -303,40 +344,168 @@ angular.module('apollo')
...
+                        environmentsNames = environmentsNames.concat(environmentName);
+                        if(($scope.selectedEnvironments).includes(environment) == false) {
+                            $scope.selectedEnvironments.push(environment);
+                        }
+                    });
+                    $scope.environmentsNamesOfStacks[environment.id] = environmentsNames;
+                }
+                else {
+                    var index = $scope.selectedEnvironments.indexOf(environment);
+                    if (index > -1) {
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: Isnt that dead code?

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 9]
```diff
@@ -303,40 +344,168 @@ angular.module('apollo')
...
+                filterGroups();
+                $scope.nextStep();
+           });
+       }
+
+       $scope.toggleSelectedServiceOrStack = function(service) {
+            var index = $scope.selectedServicesAndStacks.indexOf(service);
             if (index > -1) {
-                $scope.selectedServices.splice(index, 1);
+                $scope.selectedServicesAndStacks.splice(index, 1);
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: Exact same comments as the environment

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 10]
```diff
@@ -303,40 +344,168 @@ angular.module('apollo')
...
+
+        function toggleSelectedService(service, index) {
+            if($scope.isGroupDeployment == false) {
                 if (service.isPartOfGroup) {
-                    $scope.selectedServices = [];
-                    deploymentSteps = ["choose-service", "choose-environment", "choose-groups", "choose-version", "confirmation"];
-                    $scope.selectedServices.push(service);
-                    $scope.isGroupDeployment = true;
+                    if($scope.selectedServices.length == 0) {
+                        $scope.selectedServices = [];
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: If you got into this "if", this line is redundant

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 11]
```diff
@@ -303,40 +344,168 @@ angular.module('apollo')
...
                     $scope.selectedServices = $scope.selectedServices.filter(function(a){return !a.isPartOfGroup});
                     $scope.selectedServices.push(service);
-                    $scope.isGroupDeployment = false;
                     deploymentSteps = ["choose-service", "choose-environment", "choose-version", "confirmation"];
                 }
             }
-            updateServicesNames();
+            else {
+                if($scope.selectedServices.length == 0) {
+                    $scope.isGroupDeployment = false;
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: What? why?
What is the purpose of the previous "if" section? I don't understand this change

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 12]
```diff
@@ -445,26 +614,113 @@ angular.module('apollo')
...
+             var tempEnvironmentsStack = {};
+		     response.data.forEach(function(environmentsStack) {
+                tempEnvironmentsStack[environmentsStack.id] = environmentsStack;
+                addStackToDisplayEnvironments(environmentsStack);
+
+                var environmentsNames = [];
+                environmentsStack.environments.forEach(function(environmentId) {
+                    var environment = allEnvironments.find(x => x.id === environmentId);
+                    var environmentName = environment.name;
+                    environmentsNames = environmentsNames.concat(environmentName);
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: Why? Isnt that enough that we add all names from all environments?
There is no way an environment could be in stack but not in allEnvironments

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 13]
```diff
@@ -445,26 +614,113 @@ angular.module('apollo')
...
+		apolloApiService.getAllServices()
+		    .then(parseAllServices)
+		    .then(() => {
+		        apolloApiService.getAllServicesStacks().then(function(response) {
+                    var tempServicesStack = {};
+                    response.data.forEach(function(servicesStack) {
+                       tempServicesStack[servicesStack.id] = servicesStack;
+                       addStackToDisplayServices(servicesStack);
+
+                       var servicesNames = [];
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: Same comment here as the environment

File: ui/app/scripts/controllers/newDeploymentCtrl.js [hunk 14]
```diff
@@ -476,10 +732,15 @@ angular.module('apollo')
 		    if (environmentId !== loadedGroupsEnvironmentId || serviceId !== loadedGroupsServiceId) {
 		        loadedGroupsEnvironmentId = environmentId;
 		        loadedGroupsServiceId = serviceId;
-                apolloApiService.getGroupsPerServiceAndEnvironment(environmentId, serviceId).then(function (response) {
-                    $scope.possibleGroups = response.data;
+                return apolloApiService.getGroupsPerServiceAndEnvironment(environmentId, serviceId).then(function (response) {
+                    response.data.forEach(function(group) {
+                        $scope.possibleGroups.push(group);
+                        $scope.possibleGroupsIds.push(group.id);
```
Comment (roiravhon) [inline ui/app/scripts/controllers/newDeploymentCtrl.js]: This is not being used. please remove this from there and the declaration

File: ui/app/views/deployments/new.html [hunk 15]
```diff
@@ -16,38 +15,48 @@ <h1 class="page-header">New Deployment</h1>
...
                     <div class="table-responsive">
-                        <table class="table table-hover" datatable="ng" dt-options="dtOptions">
+                        <table class="table table-hover" datatable="ng" dt-options="dtOptions" >
                             <thead>
                             <tr>
+                                <th style="display: none;"></th>
                                 <th style="width: 20px"></th>
                                 <th style="width: 20px"></th>
-                                <th>Name</th>
+                                <th style="width: 200px">Name</th>
```
Comment (roiravhon) [inline ui/app/views/deployments/new.html]: Sure?

File: ui/app/views/deployments/new.html [hunk 16]
```diff
@@ -68,17 +76,22 @@ <h1 class="page-header">New Deployment</h1>
                         <table class="table table-hover" datatable="ng" dt-options="dtOptions">
                             <thead>
                             <tr>
+                                <th style="display: none;"></th>
                                 <th style="width: 20px"></th>
                                 <th style="width: 20px"></th>
-                                <th>Name</th>
+                                <th style="width: 200px">Name</th>
```
Comment (roiravhon) [inline ui/app/views/deployments/new.html]: Sure?

File: ui/app/scripts/controllers/newDeploymentCtrl.js (hunk 7 above)
Comment (mayhuberman) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: that way I'm not adding twice the same environment

File: ui/app/scripts/controllers/newDeploymentCtrl.js (hunk 12 above)
Comment (mayhuberman) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: I'm saving the environments names of every stack

File: ui/app/scripts/controllers/newDeploymentCtrl.js (hunk 13 above)
Comment (mayhuberman) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: what?

File: ui/app/scripts/controllers/newDeploymentCtrl.js (hunk 14 above)
Comment (mayhuberman) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: it's in use. without it, it won't know the possible groups of the selected environments.

Comment (roiravhon) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: Where? I searched your entire PR. Its just being declared, and assigned here. No one is looking on that on the JS or HTML side

File: ui/app/scripts/controllers/newDeploymentCtrl.js (hunk 13 above)
Comment (roiravhon) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: It's the exact same code.. all comments on the previous environment section applies here

File: ui/app/scripts/controllers/newDeploymentCtrl.js (hunk 12 above)
Comment (roiravhon) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: But you are doing that for all environments... It wont change in stacks

File: ui/app/scripts/controllers/newDeploymentCtrl.js (hunk 7 above)
Comment (roiravhon) [reply ui/app/scripts/controllers/newDeploymentCtrl.js]: I still don't follow. This is the addition toggle flow, and you are protecting dups in line 375, no?

**Pull request merged**